rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,752 @@
+    # bits and boolean values which indicate which values need to be removed
+    # from which bit positions to avoid covering up lower entries. If the
+    # example above were the only covered entry then ``bits_and_vals`` would
+    # contain ``(m, True)`` to indicate that all entries containing Xs or 1s in
+    # the left-most bit of interest could be removed to avoid the covered entry
+    # and ``(n, False)`` to indicate that all entries containing Xs or 0s in
+    # the right-most bit of interest could be removed to avoid covering the
+    # entry.
+    #
+    # NOTE:
+    #   ``bits_and_vals`` consists of a set of options (e.g., we *could* remove
+    #   all entries with Xs or 1s in bit ``m`` *or* we could remove all entries
+    #   with Xs or 0s in bit ``n``, either would resolve the above covering).
+    #
+    # To determine which course of action to take we build a dictionary mapping
+    # each of the pairs in ``bits_and_vals`` to the entries that would need to
+    # be removed to "set" that bit in the merged entry. For example, we might
+    # end up with:
+    #
+    #     options = {(m, True): {1, 4, 5},
+    #                (n, False): {3, 7}}
+    #
+    # Indicating that we'd need to remove entries 1, 4 and 5 from the merge to
+    # "set" the mth bit of the merged to 0 or that we'd need to remove entries
+    # 3 and 7 to set the nth bit of the merged entry to set the nth bit to 1.
+    #
+    # NOTE:
+    #   The boolean part of the pair indicates which value needs to be removed
+    #   (True -> remove all 1s; False -> remove all 0s). If all Xs and 1s in a
+    #   given bit position are removed from a merge then the merged entry is
+    #   guaranteed to have a 0 in the bit position. Vice-versa removing all Xs
+    #   and 0s in a given bit position from a merge will result in a merged
+    #   entry with a 1 in that position.
+    #
+    # As we want to make our merges as large as possible we select the smallest
+    # set of entries to remove from the merge from ``options``. After removing
+    # these entries we have a new merge with lower generality.
+    #
+    # For two reasons:
+    #   * we ignored covered entries with more ``settable`` bits there may
+    #     still be covered entries below the merged entry,
+    #   * after removing entries from the merge the merged entry is of lower
+    #     generality and is therefore nearer the top of the table so new
+    #     entries may be have become covered,
+    # we look at the merge again to see if there are any covered entries. If
+    # there are then we resolve the covering as above, if there aren't then (as
+    # far as the down-check is concerned) we have a valid merge so we return
+    # the merge.
+
+    # Set of bit positions
+    all_bits = tuple(1 << i for i in range(32))
+
+    # While the merge is still worth considering continue to perform the
+    # down-check.
+    while merge.goodness > min_goodness:
```

Comment:
If the merge goodness drops below min_goodness this function should return an empty merge (as per the docstring). I can't see any code which would cause this to happen. (I *can* see code for returning empty merges when there are no settable bits, though.)